peer/upstream/upstream.go

```diff
@@ -0,0 +1,353 @@
+// Copyright (c) 2014-2019 Bitmark Inc.
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package upstream
+
+import (
+	"encoding/binary"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/bitmark-inc/bitmarkd/blockdigest"
+	"github.com/bitmark-inc/bitmarkd/fault"
+	"github.com/bitmark-inc/bitmarkd/mode"
+	"github.com/bitmark-inc/bitmarkd/util"
+	"github.com/bitmark-inc/bitmarkd/zmqutil"
+	"github.com/bitmark-inc/logger"
+	zmq "github.com/pebbe/zmq4"
+)
+
+type UpstreamIntf interface {
+	ActiveInPastSeconds(time.Duration) bool
+	Connect(*util.Connection, []byte) error
+	ConnectedTo() *zmqutil.Connected
+	Destroy()
+	IsConnectedTo([]byte) bool
+	IsConnected() bool
+	GetBlockData(uint64) ([]byte, error)
+	GetClient() zmqutil.ClientIntf
+	RemoteAddr() (string, error)
+	LocalHeight() uint64
+	RemoteDigestOfHeight(uint64) (blockdigest.Digest, error)
+	RemoteHeight() (uint64, error)
+	CachedRemoteHeight() uint64
+	CachedRemoteDigestOfLocalHeight() blockdigest.Digest
+	Name() string
+	ResetServer()
+}
+
+// Upstream - structure to hold an upstream connection
+type Upstream struct {
+	sync.RWMutex
+
+	UpstreamIntf
+
+	log                       *logger.L
+	name                      string
+	client                    zmqutil.ClientIntf
+	connected                 bool
+	remoteHeight              uint64
+	localHeight               uint64
```

Comment:
upstream meaning basically is a server that this node connected to. I don't understand why do you define different height variable `remoteHeight` and `localHeight`? It should has only `height` represents the block height of that server. Please explain more.

Reply:
I think this is not suggestion, but you are unclear about what task try to do, so I suggest you read the task description on Monday

Reply:
I am focusing on the code. Base on purpose of upstream, at least in my understanding, i question you about the 2 variables since it concerned me. Please explain to me.

Reply:
please try to understand first? it's code review not design review

Reply:
@hxw How do you feel?